compiler/semantic/checker.go

```diff
@@ -0,0 +1,1081 @@
+			}
+			return c.arithmetic(e.LHS, e.RHS, lhs, rhs)
+		default:
+			panic(e.Op)
+		}
+	case *sem.CallExpr:
+		var types []super.Type
+		for _, e := range e.Args {
+			types = append(types, c.expr(typ, e))
+		}
+		if isBuiltin(e.Tag) {
+			return c.callBuiltin(e, types)
+		}
+		return c.callFunc(e, types)
+	case *sem.CondExpr:
+		c.boolean(e.Cond, c.expr(typ, e.Cond))
+		return c.fuse([]super.Type{c.expr(typ, e.Then), c.expr(typ, e.Else)})
+	case *sem.DotExpr:
+		typ, _ := c.deref(e.Node, c.expr(typ, e.LHS), e.RHS)
+		return typ
+	case *sem.IndexExpr:
+		typ, _ := c.indexOf(e.Expr, e.Index, c.expr(typ, e.Expr), c.expr(typ, e.Index))
+		return typ
+	case *sem.IsNullExpr:
+		c.expr(typ, e.Expr)
+		return super.TypeBool
+	case *sem.LiteralExpr:
+		if val, err := sup.ParseValue(c.t.sctx, e.Value); err == nil {
+			return val.Type()
+		}
+		return c.unknown
+	case *sem.MapCallExpr:
+		containerType := c.expr(typ, e.Expr)
+		elemType := c.isContainer(containerType)
+		if elemType == nil {
+			c.error(e.Expr, errors.New("map entity must be an array or set"))
+			return c.unknown
+		}
+		c.epush()
+		lambdaType := c.expr(elemType, e.Lambda)
+		errs := c.epop()
+		if len(errs) != 0 {
+			c.error(errs[0].loc, fmt.Errorf("in functon called from map: %w", errs[0].err))
+		}
+		return lambdaType
+	case *sem.MapExpr:
+		// fuser could take type at a time instead of array
+		var keyTypes []super.Type
+		var valTypes []super.Type
+		for _, entry := range e.Entries {
+			keyTypes = append(keyTypes, c.expr(typ, entry.Key))
+			valTypes = append(valTypes, c.expr(typ, entry.Value))
+		}
+		return c.t.sctx.LookupTypeMap(c.fuse(keyTypes), c.fuse(valTypes))
+	case *sem.RecordExpr:
+		return c.recordElems(typ, e.Elems)
+	case *sem.RegexpMatchExpr:
+		if !hasString(c.expr(typ, e.Expr)) {
+			c.error(e.Expr, errors.New("string match must apply to type string"))
+		}
+		return super.TypeBool
+	case *sem.RegexpSearchExpr:
+		if !hasString(c.expr(typ, e.Expr)) {
+			c.error(e.Expr, errors.New("string match must apply to type string"))
+		}
+		return super.TypeBool
+	case *sem.SearchTermExpr:
+		c.expr(typ, e.Expr)
+		return super.TypeBool
+	case *sem.SetExpr:
+		return c.t.sctx.LookupTypeArray(c.arrayElems(typ, e.Elems))
+	case *sem.SliceExpr:
+		c.integer(e.From, c.expr(typ, e.From))
+		c.integer(e.To, c.expr(typ, e.To))
+		container := c.expr(typ, e.Expr)
+		c.sliceable(e.Expr, container)
+		return container
+	case *sem.SubqueryExpr:
+		typ = c.seq(typ, e.Body)
+		if e.Array {
+			typ = c.t.sctx.LookupTypeArray(typ)
+		}
+		return typ
+	case *sem.ThisExpr:
+		for _, field := range e.Path {
+			typ, _ = c.deref(e.Node, typ, field)
+		}
+		return typ
+	case *sem.UnaryExpr:
+		typ = c.expr(typ, e.Operand)
+		switch e.Op {
+		case "-":
+			c.number(e.Operand, typ)
+			return typ
+		case "!":
+			c.boolean(e, typ)
+			return super.TypeBool
+		default:
+			panic(e.Op)
+		}
+	default:
+		panic(e)
+	}
+}
+
+func (c *checker) isContainer(containerType super.Type) super.Type {
```

Comment:
Nit: It's more idiomatic to add a bool as a second return value to signal that a predicate has evaluated to false instead of signaling false by returning nil.